mantidproject/mantid
# FindEPP rewrite in c++

`FindEPP` is used when finding elastic TOF channel. The existing algorithm was written in python and has been found to be very slow for an instrument like IN5 at ILL, where one has 10ˆ5 spectra. This PR provides C++ version of the same algorithm, implemented as version 2. It is now 10 to 50 times faster (dependent on number of cores and whether OpenMP is enabled), while providing identical results. A tiny difference is introduced in the behaviour of the algorithm concerning the cases when the fit fails (as described in the documentation), however this should be transparent.

**To test:**
1. Run this script and make sure the answer is positive.
```
ws = CreateSampleWorkspace(Function="User Defined", UserDefinedFunction="name=LinearBackground, A0=0.3, A1=0.001;name=Gaussian, PeakCentre=6000, Height=5, Sigma=75", NumBanks=10,BankPixelWidth=10, XMin=4005.75, XMax=7995.75, BinWidth=2.1, Random=True)
new = FindEPP(ws)
old = FindEPP(ws, Version=1)
new_peaks = ConvertTableToMatrixWorkspace(new,ColumnX="WorkspaceIndex",ColumnY="PeakCentre", ColumnE="PeakCentreError")
old_peaks = ConvertTableToMatrixWorkspace(old,ColumnX="WorkspaceIndex",ColumnY="PeakCentre", ColumnE="PeakCentreError")
result = CompareWorkspaces(new_peaks,old_peaks,Tolerance=0.00001)
print result[0]
```
2. Open the histories of new and old table workspaces, and compare the execution times.
3. Optionally, load the attached IN5 raw data, and run FindEPP version 1 and 2 on a real life example. Report the execution times. If you have OpenMP enabled, follow that the loop over the spectra in the  new version of the algorithm is indeed run in parallel.

[106948.tar.zip](https://github.com/mantidproject/mantid/files/913582/106948.tar.zip)



Fixes #19133 .



---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
